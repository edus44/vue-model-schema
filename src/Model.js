const isPlainObject = require('lodash/isPlainObject')
const { getValidationsFromModel } = require('./validations')

class Model {
  constructor(name, fieldsDefs) {
    if (typeof name !== 'string') {
      fieldsDefs = name
      name = 'default'
    }
    this.name = name
    this.fields = {}
    this.identifier = 'id'
    this.defaultSort = 'createdAt'
    this.listable = []
    this.sortable = []
    this.filtrable = []

    for (const fieldName in fieldsDefs) {
      const field = normalizeField(fieldName, fieldsDefs[fieldName])
      if (field.identifier) this.identifier = fieldName
      if (field.defaultSort) this.defaultSort = fieldName
      if (field.listable) this.listable.push(field)
      if (field.sortable) this.sortable.push(field)
      if (field.filtrable) this.filtrable.push(field)

      this.fields[fieldName] = field
    }
    Object.freeze(this)
  }

  getField(fieldName) {
    const levels = Array.isArray(fieldName) ? fieldName : fieldName.split('.')
    const firstLevel = this.fields[levels[0]]
    if (levels.length > 1 && firstLevel && firstLevel.type instanceof Model) {
      return firstLevel.type.getField(levels.slice(1))
    } else {
      return firstLevel
    }
  }

  ns(fieldName) {
    return `${this.name}${fieldName ? '.' + fieldName : ''}`
  }

  getValidations(params, modifiers) {
    return getValidationsFromModel(this, params, modifiers)
  }
}

/**
 * Normalize field definition
 */
function normalizeField(name, def) {
  let field = def
  if (!isPlainObject(field)) field = { type: field }

  if (Array.isArray(field.type)) {
    field.type = field.type[0]
    field.isArray = true
  }

  if (!field.type) field.type = null

  field.name = name

  return Object.freeze(field)
}

module.exports = Model
