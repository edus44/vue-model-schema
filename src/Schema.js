/* eslint-disable new-cap */
import TYPES from './types'
import sortBy from 'lodash.sortby'
import extendWith from 'lodash.assigninwith'
import extend from 'lodash.assignin'
import { getValidationsFromModel } from './validationHelpers'
import moment from 'moment'

export default class Schema {
  /**
   * Initialize every field
   * @param  {Object} obj
   * @return {this}
   */
  constructor(obj) {
    // Shim fields store
    if (!this.constructor._fields) {
      this.constructor._fields = {}
    }

    Object.defineProperty(this, '$', {
      value: obj,
    })
    this.$setObject(obj)
  }

  /**
   * Identify itself
   * @type {Boolean}
   */
  @hidden
  static _isSchema = true

  /**
   * Return all fieldNames sorted by order property (inherited also)
   * @return {Array<String>}
   */
  static getFieldNames() {
    if (!Object.prototype.hasOwnProperty.call(this, '_fieldNames')) {
      const tmp = []
      for (const fieldName in this._fields) tmp.push(fieldName)

      this._fieldNames = sortBy(tmp, x => this._fields[x].order || 0)
    }
    return this._fieldNames
  }

  static getFields() {
    return this._fields
  }

  static getField(name) {
    const levels = name.split('.')
    const firstLevel = this._fields[levels[0]]
    if (firstLevel && firstLevel.type._isSchema) {
      return firstLevel.type.getField(levels.slice(1).join('.'))
    } else {
      return firstLevel
    }
  }

  static getFieldsInfo() {
    if (!Object.prototype.hasOwnProperty.call(this, '_fieldsInfo')) {
      const fieldNames = this._fieldNames || this.getFieldNames()
      this._fieldsInfo = fieldNames.reduce(
        (acc, fieldName) => {
          const field = this._fields[fieldName]
          if (field.filtrable) acc.filtrable.push(field)
          if (field.sortable) acc.sortable.push(field)
          if (field.listable) acc.listable.push(field)
          return acc
        },
        {
          sortable: [],
          filtrable: [],
          listable: [],
        }
      )
      Object.freeze(this._fieldsInfo)
    }
    return this._fieldsInfo
  }

  /**
   * As getFieldsInfo but with optional fields to extend
   * over own fields
   * Creates a virtual model merging own an extended fields to
   * then sort and returns info
   * @param  {Object} extendFields Same format as _fields
   * @return {Object} Like getFieldsInfo
   */
  static getFieldsInfoExtended(extendFields) {
    if (typeof extendFields !== 'object') return this.getFieldsInfo()

    const virtualModel = {
      _fields: extendWith({}, this._fields, extendFields, (a, b, name) => {
        return extend({ name }, a, b)
      }),
    }
    this.getFieldNames.call(virtualModel)
    return this.getFieldsInfo.call(virtualModel)
  }

  static get namespace() {
    return this._resource.namespace
  }

  static ns(fieldName) {
    fieldName = fieldName ? '.' + fieldName : ''
    return `${this.namespace}${fieldName}`
  }

  /**
   * Clone or create a new instance
   * @param  {Object} obj Optional
   * @return {Object} Instance of itself
   */
  static from(obj) {
    if (obj instanceof this) return obj.$clone()
    else return new this(obj)
  }

  static getValidations(fieldNames) {
    return getValidationsFromModel(this, fieldNames)
  }

  /**
   * Set a key and cast its value
   * @param {String} key
   * @param {Any} val
   * @return {this}
   */
  $set(key, val) {
    const field = this.constructor._fields[key]
    if (!field) throw new Error(`Field ${key} does not exist in ${this.constructor.name}`)

    if (field.isArray) {
      this[key] = mapArray(val, x => cast(x, field))
    } else {
      this[key] = cast(val, field)
    }
    return this
  }

  /**
   * Set a property as undefined
   * @param  {String} key
   * @return {this}
   */
  $unset(key) {
    delete this[key]
    return this
  }

  /**
   * Add a value to a array field
   * @param {String} key isArray Field
   * @param {Any} val
   * @return {this}
   */
  $add(key, val) {
    const field = this.constructor._fields[key]
    if (!field) throw new Error(`Field ${key} does not exist in ${this.constructor.name}`)
    if (!field.isArray) throw new Error(`Field ${key} in ${this.constructor.name} is not array`)
    this[key].push(cast(val, field))
    return this
  }

  /**
   * Cast every value through its field type
   * @return {this}
   */
  $cast() {
    this.$eachField(name => {
      this.$set(name, this[name])
    })
    return this
  }

  /**
   * Assign values recursively
   * overriding every one
   * @param {Object} obj
   * @return {this}
   */
  $setObject(obj = {}) {
    // Initalize field values
    this.$eachField(name => {
      this.$set(name, obj[name])
    })

    return this
  }

  /**
   * Export schema content as object
   * @return {Object}
   */
  $toObject(opts) {
    function getValue(val, field) {
      if (field.type._isSchema) {
        // Schema fields that can be nullable
        if (val === null) {
          return null
        } else if (typeof val.$toObject !== 'function') {
          // In case a field type Schema is nullable an the value was null, it will be a plain object
          return JSON.parse(JSON.stringify(val))
        } else {
          return val.$toObject()
        }
      } else if (moment.isMoment(val)) {
        return val.toISOString()
      } else {
        return cast(val, field)
      }
    }

    const obj = {}
    this.$eachField((name, field) => {
      const val = this[name]
      if (field.isArray) {
        obj[name] = mapArray(val, x => getValue(x, field))
      } else {
        obj[name] = getValue(val, field)
      }
    }, opts)

    if (typeof this.constructor.$transform === 'function')
      return this.constructor.$transform(obj, opts)

    return obj
  }

  /**
   * Iterate over fields
   * @param  {Function} fn Arguments: name,definition,isSchema
   * @return {this}
   */
  $eachField(fn, opts) {
    // Get constructor function
    const ctor = this.constructor
    const virtuals = opts && opts.virtuals === true

    // Iterate over fields
    for (const fieldName in ctor._fields) {
      // Get field definition
      const field = ctor._fields[fieldName]
      if (field.virtual && !virtuals) continue
      fn.call(this, fieldName, field)
    }
    return this
  }

  $clone() {
    return new this.constructor(this.$toObject())
  }

  $id() {
    return this[this.constructor._resource.rowId]
  }

  $isNew() {
    return !this.$id()
  }

  $getSchema() {
    return this.constructor
  }
}

/**
 * Decorator to set a property as
 * non writable nor enumerable
 */
function hidden(target, name, descriptor) {
  descriptor.writable = false
  descriptor.enumerable = false
  return descriptor
}

/**
 * Used to cast a value taking into account
 * its type and definition
 * @param  {Any} value
 * @param  {Object} field
 * @return {Any}
 */
function cast(value, field) {
  // Check if type is another schema
  if (field.type._isSchema) {
    if (field.nullable && value === null) {
      // it wont be a Schema anymore
      return null
    }
    return new field.type(value)
  }

  // Get type
  const type = TYPES[field.type.name]

  // Apply default value
  if (field.default !== undefined && value === undefined) {
    value = typeof field.default === 'function' ? field.default() : field.default
  }

  // Check if need casting
  if (value !== undefined && !type.is(value)) {
    if (field.nullable && value === null) return null
    value = type.cast(value)
  }

  return value
}

function mapArray(arr, fn) {
  return Array.isArray(arr) ? arr.map(fn) : []
}
