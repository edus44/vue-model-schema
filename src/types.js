import isPlainObject from 'lodash.isplainobject'
import moment from 'moment'

export const Any = { name: 'Any' }

export default {
  Any: {
    is: () => true,
    // cast: x => x,
  },
  Boolean: {
    is: x => x === true || x === false,
    cast: Boolean,
  },
  Array: {
    is: Array.isArray,
    cast: x => (x ? Array.from(x) : []),
  },
  String: {
    is: x => typeof x === 'string',
    cast: str => (str ? String(str) : ''),
  },
  Number: {
    is: x => typeof x === 'number',
    cast: Number,
  },
  Object: {
    is: () => false,
    cast: x => (isPlainObject(x) ? JSON.parse(JSON.stringify(x)) : {}),
  },
  Date: {
    is: x => moment.isMoment(x),
    cast(x) {
      if (typeof x === 'number') {
        return moment(x)
      } else {
        return moment(x, moment.ISO_8601)
      }
    },
  },
}
