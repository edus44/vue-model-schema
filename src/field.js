const { Any, TYPES } = require('./types')

/**
 * Normalize field definition
 * @param  {Function} initializer Descriptor initializer
 * @return {Object} Field
 */
function normalizeField(name, definition, ext) {
  let field = definition || { type: Any }

  if (Array.isArray(field)) {
    return normalizeField(name, field[0], { isArray: true })
  }

  // If value is type, wrap it
  if (typeof field == 'function') field = { type: field }

  // Default type to Any
  if (!field.type) field.type = Any

  field.name = name

  return Object.freeze(Object.assign(field, ext))
}

/**
 * Used to cast a value taking into account
 * its type and definition
 * @param  {Any} value
 * @param  {Object} field
 * @return {Any}
 */
function cast(value, field) {
  // Check if type is another schema
  if (field.type._isSchema) {
    if (field.nullable && value === null) {
      // it wont be a Schema anymore
      return null
    }
    return new field.type(value)
  }

  // Get type
  const type = TYPES[field.type.name]

  // Apply default value
  if (field.default !== undefined && value === undefined) {
    value = typeof field.default === 'function' ? field.default() : field.default
  }

  // Check if need casting
  if (value !== undefined && !type.is(value)) {
    if (field.nullable && value === null) return null
    value = type.cast(value)
  }

  return value
}

function mapArray(arr, fn) {
  return Array.isArray(arr) ? arr.map(fn) : []
}

function isMoment(obj) {
  return obj != null && obj._isAMomentObject != null
}

function getValue(val, field) {
  if (field.type._isSchema) {
    // Schema fields that can be nullable
    if (val === null) {
      return null
    } else if (typeof val.$toObject !== 'function') {
      // In case a field type Schema is nullable an the value was null, it will be a plain object
      return JSON.parse(JSON.stringify(val))
    } else {
      return val.$toObject()
    }
  } else if (isMoment(val)) {
    return val.toISOString()
  } else {
    return cast(val, field)
  }
}

module.exports = { normalizeField, cast, mapArray, getValue }
