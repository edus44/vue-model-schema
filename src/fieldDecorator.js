import { Any } from './types'

/**
 * Field properties
 * 
    type: Function,
    sortable: Boolean || Object,
    listable: Boolean || Object,
    filtrable: Boolean || Object,
    validations: Object,
    order: Number
 */

/**
 * Normalize field definition
 * @param  {Function} initializer Descriptor initializer
 * @return {Object} Field
 */
function normalizeField(name, initializer, ext) {
  let field
  // If not initial value return default def
  if (typeof initializer === 'function') field = initializer()
  else field = { type: Any }

  if (Array.isArray(field)) {
    const arrayField = normalizeField(name, () => field[0], { isArray: true })
    return arrayField
  }

  // If value is type, wrap it
  if (typeof field !== 'object') field = { type: field }

  // Default type to Any
  if (!field.type) field.type = Any

  field.name = name

  return Object.freeze(Object.assign(field, ext))
}

/**
 * Field decorator for Model classes
 * @param  {Function} target
 * @param  {String} name
 * @param  {Object} descriptor
 * @return {Object}
 */
export function fieldDecorator(target, name, descriptor) {
  const ctor = target.constructor

  // Initialize fields store
  // Support inheritance
  if (!Object.prototype.hasOwnProperty.call(ctor, '_fields')) {
    ctor._fields = Object.create(ctor._fields || null)
  }

  // Save field definition
  ctor._fields[name] = normalizeField(name, descriptor.initializer)

  return {
    configurable: false,
    enumerable: true,
    writable: true,
  }
}

/**
 * Virtual field decorator for Model classes
 * @param  {Function} target
 * @param  {String} name
 * @param  {Object} descriptor
 * @return {Object}
 */
export function virtualFieldDecorator(target, name, descriptor) {
  const ctor = target.constructor

  // Initialize fields store
  // Support inheritance
  if (!Object.prototype.hasOwnProperty.call(ctor, '_fields')) {
    ctor._fields = Object.create(ctor._fields || null)
  }

  // Save field definition
  ctor._fields[name] = normalizeField(name, descriptor.initializer, { virtual: true })

  const get = typeof ctor._fields[name].get === 'function' ? ctor._fields[name].get : () => {}
  const set = typeof ctor._fields[name].set === 'function' ? ctor._fields[name].set : () => {}

  return {
    configurable: false,
    enumerable: true,
    get,
    set,
  }
}
