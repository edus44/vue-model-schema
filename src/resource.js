const merge = require('lodash/merge')

function resource({ client, actions = {}, methods = {}, config = {} }) {
  Object.assign(actions, resource.configuration.actions)
  Object.assign(methods, resource.configuration.methods)

  for (const actionName in actions) {
    methods[actionName] = inputConfig => {
      return request(inputConfig, actions[actionName], config)
    }
  }

  return methods
}

resource.configuration = {
  methods: {
    findOne() {},
  },
  actions: {
    list: {
      method: 'get',
    },
    detail: {
      method: 'get',
      prependUrl: '/:id',
    },
    create: {
      method: 'post',
    },
    update: {
      method: 'put',
      prependUrl: '/:id',
    },
    delete: {
      method: 'delete',
      prependUrl: '/:id',
    },
    multiDelete: {
      method: 'delete',
    },
  },
}

function request(inputConfig, actionConfig, resourceConfig) {
  // Merge configs, sort by precedence
  const config = merge({}, resourceConfig, actionConfig, inputConfig)

  // Compose url
  const fullUrl = `${config.appendUrl || ''}${config.url || ''}${config.prependUrl || ''}`
  config.url = interpolate(fullUrl, config.urlParams)
  console.log(config)
  return config
}

function interpolate(str, values) {
  for (const key in values) str = str.replace(`:${key}`, encodeURIComponent(values[key]))
  return str
}

module.exports = resource
