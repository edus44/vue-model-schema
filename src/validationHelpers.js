import set from 'lodash.set'
import * as validators from 'vuelidate/lib/validators'

export function getValidationsFromModel(model, params) {
  const v = {}
  const modifiers = typeof params === 'object' ? params : {}
  const fieldNames = Array.isArray(params) ? params : Object.keys(params)
  // Iterate fields
  for (const fieldName of fieldNames) {
    const fieldValidations = getFieldValidations(model, fieldName, modifiers[fieldName], this)
    if (fieldValidations) set(v, fieldName, fieldValidations)
  }

  return v
}

function getFieldValidations(model, fieldName, modifier = {}, vm) {
  // Get validations from field definition
  const field = model.getField(fieldName)
  const validations = field && typeof field.validations === 'object' && field.validations
  if (!validations) return
  const v = {}

  // Iterate over field validators
  for (const name in validations) {
    const args = modifier[name] || validations[name]
    const validator = getValidator(name, args)
    if (validator) v[name] = validator
  }
  return v
}

/**
 * Get vuelidate validator
 * @param  {String} name
 * @param  {Any} args If function return itself
 *                    If array apply to found validator
 *                    If truthy return found validator
 *                    If falsey return undefined
 * @return {Function} Validator function or undefined
 */
function getValidator(name, args) {
  if (!args) return

  if (typeof args === 'function') return args

  const fn = validators[name]

  if (!fn) return

  if (Array.isArray(args)) {
    return fn(...args)
  }

  return fn
}
