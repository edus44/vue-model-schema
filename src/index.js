import { fieldDecorator as field, virtualFieldDecorator as virtual } from './fieldDecorator'

import Schema from './Schema'

export { field, virtual, Schema }
