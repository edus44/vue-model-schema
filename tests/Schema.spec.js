/* eslint-env jest */
import { field, virtual, Schema } from '../src'

describe('Schema class', () => {
  it('should be a function', () => {
    expect(Schema).toEqual(expect.any(Function))
  })

  it('should have hidden identify static property', () => {
    const descriptor = Object.getOwnPropertyDescriptor(Schema, '_isSchema')
    expect(descriptor.enumerable).toEqual(false)
    expect(descriptor.writable).toEqual(false)
    expect(descriptor.value).toEqual(true)
  })
})

describe('Schema normal field instantiation', () => {
  it('should evaluate', () => {
    class Foo extends Schema {
      @field name = String
    }
    expect(Foo._fields.name).toHaveProperty('type', String)
  })

  it('should instance with default value', () => {
    class Foo extends Schema {
      @field name = { default: 'Bob' }
    }
    const foo = new Foo()
    expect(foo).toBeInstanceOf(Foo)
    expect(foo.name).toBe('Bob')
  })

  it('should instance with value', () => {
    class Foo extends Schema {
      @field name = { default: 'Bob' }
    }
    const foo = new Foo({ name: 'Alice' })
    expect(foo.name).toBe('Alice')
  })

  it('should change a value', () => {
    class Foo extends Schema {
      @field name
    }
    const foo = new Foo({ name: 'Alice' })
    foo.$set('name', 'Bob')
    expect(foo.name).toBe('Bob')
  })

  it('should fail setting a non-exist key', () => {
    class Foo extends Schema {}
    const foo = new Foo()
    expect(() => foo.$set('name', 'Bob')).toThrow('Field name does not exist in Foo')
  })

  it('should instance without fields', () => {
    class Foo extends Schema {}
    const foo = new Foo({ name: 'Alice' })
    expect(foo.name).toBeUndefined()
  })

  it('should export to object', () => {
    class Address extends Schema {
      @field city
    }
    class Foo extends Schema {
      @field name
      @field address = Address
    }
    const foo = new Foo({ name: 'Bob', address: { city: 'NY' } })
    const obj = foo.$toObject()

    expect(obj).toHaveProperty('name', 'Bob')
    expect(obj).toHaveProperty('address.city', 'NY')
    expect(foo.address).toBeInstanceOf(Address)
  })

  it('should return fields', async () => {
    class Foo extends Schema {
      @field name
    }
    expect(Foo.getFields()).toBe(Foo._fields)
  })

  it('should return fields from a string path', async () => {
    class Foo extends Schema {
      @field name
      @field properties = class Properties extends Schema {
        @field surname
      }
    }
    expect(Foo.getField('name')).toBe(Foo._fields.name)
    expect(Foo.getField('properties.surname')).toBe(Foo._fields.properties.type._fields.surname)
  })

  it.skip('should return namespace', async () => {
    @resource({
      name: 'Foo',
    })
    class Foo extends Schema {}
    expect(Foo.ns()).toBe('model.Foo')
    expect(Foo.ns('name')).toBe('model.Foo.name')
  })

  it('should clone object', async () => {
    class Foo extends Schema {
      @field name
    }
    const foo = new Foo({ name: 'alice' })
    const cloned = Foo.from(foo)

    expect(cloned).not.toBe(foo)
    expect(cloned.$toObject()).toEqual(foo.$toObject())
  })

  it('should create from object', async () => {
    class Foo extends Schema {
      @field name
    }
    const foo = Foo.from({ name: 'alice' })

    expect(foo).toBeInstanceOf(Foo)
    expect(foo.name).toBe('alice')
  })
})

describe('Fields properties', () => {
  it('should save fields properties', async () => {
    const def = {
      type: String,
      sortable: true,
      listable: true,
      filtrable: true,
      validations: {},
      order: 100,
    }
    class Foo extends Schema {
      @field name = def
    }
    expect(Foo._fields.name).toMatchObject(def)
  })

  it('should collect fields info', async () => {
    class Foo extends Schema {
      @field a = { sortable: true }
      @field b = { listable: true }
      @field c = { filtrable: true }
    }
    expect(Foo.getFieldsInfo()).toMatchSnapshot()
  })

  it('should extend fields info', async () => {
    class Foo extends Schema {
      @field a = { sortable: true }
      @field b = { listable: true }
      @field c = { filtrable: true }
    }
    expect(
      Foo.getFieldsInfoExtended({
        d: { listable: true },
        c: { filtrable: false },
      })
    ).toMatchSnapshot()
  })
})

describe('Schema basic instance methods', () => {
  it.skip('should return id', async () => {
    @resource({
      name: 'Foo',
      rowId: 'myId',
    })
    class Foo extends Schema {
      @field myId
    }
    const foo = new Foo({ myId: 'bar' })
    expect(foo.$id()).toBe('bar')
  })

  it.skip('should if is new', async () => {
    @resource({
      name: 'Foo',
      rowId: 'myId',
    })
    class Foo extends Schema {
      @field myId
    }
    const foo1 = new Foo({ myId: 'bar' })
    const foo2 = new Foo({})
    expect(foo1.$isNew()).toBe(false)
    expect(foo2.$isNew()).toBe(true)
  })

  it('should return own schema', async () => {
    class Foo extends Schema {}
    const foo = new Foo()
    expect(foo.$getSchema()).toBe(Foo)
  })
})

describe('Schema types', () => {
  it('Boolean', () => {
    class Foo extends Schema {
      @field a = Boolean
      @field b = Boolean
    }

    // Initial casting
    const foo = new Foo({ a: 0, b: '0' })
    expect(foo.a).toEqual(false)
    expect(foo.b).toEqual(true)

    // Using setters
    foo.$set('a', 1)
    foo.$unset('b')
    expect(foo.a).toEqual(true)
    expect(foo.b).toBeUndefined()

    // Classic
    delete foo.a
    foo.b = 0
    const fooObj = foo.$toObject()
    expect(foo.b).toEqual(0)
    expect(fooObj.a).toBeUndefined()
    expect(fooObj.b).toEqual(false)
  })

  it('Array', () => {
    class Foo extends Schema {
      @field a = Array
    }
    const foo = new Foo({
      a: 0,
    })
    expect(foo.a).toEqual(expect.any(Array))
    expect(foo.a).toHaveLength(0)

    foo.$set('a', '123')
    expect(foo.a).toEqual(expect.any(Array))
    expect(foo.a).toHaveLength(3)
  })

  it('String', () => {
    class Foo extends Schema {
      @field a = String
    }
    const foo = new Foo({
      a: { foo: 1 },
    })
    expect(foo.a).toEqual('[object Object]')
    foo.$set('a', 2)
    expect(foo.a).toEqual('2')
  })

  it('Number', () => {
    class Foo extends Schema {
      @field a = Number
    }
    const foo = new Foo({
      a: { foo: 1 },
    })
    expect(foo.a).toEqual(NaN)
    foo.$set('a', '2')
    expect(foo.a).toEqual(2)
    foo.$set('a', null)
    expect(foo.a).toEqual(0)
    foo.$set('a', NaN)
    expect(foo.a).toEqual(NaN)
    foo.$set('a', '2.50')
    expect(foo.a).toEqual(2.5)
  })

  it('Object', () => {
    class Foo extends Schema {
      @field a = Object
    }
    const foo = new Foo({
      a: null,
    })
    expect(foo.a).toEqual({})
    foo.$set('a', '2')
    expect(foo.a).toEqual({})
    foo.$set('a', { bar: 1 })
    expect(foo.a).toHaveProperty('bar', 1)
    foo.$set('a', [1, 2, 3])
    expect(foo.a).toEqual({})
  })

  it('Date', () => {
    const iso = '2017-11-13T19:16:51+01:00'
    const ts = 1513610527293
    class Foo extends Schema {
      @field a = Date
      @field b = Date
    }
    const foo = new Foo({
      a: iso,
      b: ts,
    })

    expect(foo.a.isValid()).toEqual(true)
    expect(foo.a.toISOString()).toEqual('2017-11-13T18:16:51.000Z')

    expect(foo.b.isValid()).toEqual(true)
    expect(foo.b.toISOString()).toEqual('2017-12-18T15:22:07.293Z')

    foo.$set('a', 'bar')
    expect(foo.a.isValid()).toEqual(false)
    expect(foo.a.toISOString()).toEqual(null)
  })
})

describe('Schema array field instantiation', () => {
  it('should evaluate', () => {
    class Foo extends Schema {
      @field names = [String]
    }
    expect(Foo._fields.names).toHaveProperty('type', String)
    expect(Foo._fields.names).toHaveProperty('isArray', true)
  })

  it('should instantiate with simple type', () => {
    class Foo extends Schema {
      @field names = [String]
    }
    const foo = new Foo({ names: ['Alice', 'Bob'] })
    expect(foo).toBeInstanceOf(Foo)
    expect(foo.names).toHaveLength(2)
    expect(foo.names[0]).toEqual('Alice')
    expect(foo.names[1]).toEqual('Bob')
  })

  it('should instantiate always as array', () => {
    class Foo extends Schema {
      @field names = [String]
      @field cities = []
    }
    const foo = new Foo({ names: 'foo', cities: { ny: true } })
    expect(foo).toBeInstanceOf(Foo)
    expect(foo.names).toEqual(expect.any(Array))
    expect(foo.names).toHaveLength(0)
    expect(foo.cities).toEqual(expect.any(Array))
    expect(foo.cities).toHaveLength(0)
  })

  it('should instantiate with Schema type', () => {
    class Name extends Schema {
      @field first
      @field second
    }
    class Foo extends Schema {
      @field names = [Name]
    }
    const foo = new Foo({ names: [{ first: 'John', second: 'Doe' }] })
    expect(foo.names[0]).toBeInstanceOf(Name)
    expect(foo.names[0]).toHaveProperty('first', 'John')
    expect(foo.names[0]).toHaveProperty('second', 'Doe')
  })

  it('should instantiate with Schema type in type property', () => {
    class Name extends Schema {
      @field first
      @field second
    }
    class Foo extends Schema {
      @field names = [
        {
          type: Name,
        },
      ]
    }
    const foo = new Foo({ names: [{ first: 'John', second: 'Doe' }] })
    expect(foo.names[0]).toBeInstanceOf(Name)
    expect(foo.names[0]).toHaveProperty('first', 'John')
    expect(foo.names[0]).toHaveProperty('second', 'Doe')
  })

  it('should export with Schema type', () => {
    class Name extends Schema {
      @field first
      @field second
    }
    class Foo extends Schema {
      @field names = [Name]
    }
    const foo = new Foo({ names: [{ first: 'John', second: 'Doe' }] })
    const obj = foo.$toObject()
    expect(obj.names[0]).toHaveProperty('first', 'John')
    expect(obj.names[0]).toHaveProperty('second', 'Doe')
  })

  it('should add a casted item to array', () => {
    class Foo extends Schema {
      @field a = [Number]
    }
    const foo = new Foo({ a: [1, '2', ['3']] })
    foo.$add('a', '4')
    expect(foo.a).toEqual([1, 2, 3, 4])
    foo.a.push({})

    const fooObj = foo.$toObject()
    expect(fooObj.a).toEqual([1, 2, 3, 4, NaN])
  })

  it('should cast values setted without set method', () => {
    class Foo extends Schema {
      @field a = Boolean
      @field b = [Number]
    }
    const foo = new Foo()
    foo.a = 1
    foo.b = ['1', ['2']]
    foo.$cast()
    expect(foo.a).toEqual(true)
    expect(foo.b).toEqual([1, 2])
  })
})

it('should check nullable schema type field', () => {
  class Name extends Schema {
    @field name
  }
  class Fullname extends Schema {
    @field first = {
      type: Name,
    }
    @field last = {
      type: Name,
      nullable: true,
    }
  }
  const fullname = new Fullname()
  fullname.first.name = 'Bat'
  fullname.last.name = null
  fullname.$cast()
  expect(fullname.first).toEqual({ name: 'Bat' })
  expect(fullname.last).toEqual({ name: null })
})

it('should break null schema field not nullable', () => {
  class Name extends Schema {
    @field name
  }
  class Fullname extends Schema {
    @field first = {
      type: Name,
    }
    @field last = {
      type: Name,
      nullable: false,
    }
  }
  const fullname = new Fullname()
  fullname.first.name = 'Bat'
  fullname.last = null
  try {
    fullname.$cast()
  } catch (error) {
    expect(error).toBeDefined()
  }
  expect(fullname.first).toEqual({ name: 'Bat' })
})

describe('Schema inheritance', () => {
  it('should inherit fields from other Schema', () => {
    class Base extends Schema {
      @field x = { type: Number, default: 0 }
      @field y = { type: Number, default: 0 }
    }
    class Cube extends Base {
      @field width = { type: Number, default: 0 }
      @field height = { type: Number, default: 0 }
    }
    const cube = new Cube({ x: 1, height: 1 })
    const cubeObj = cube.$toObject()
    const expected = { width: 0, height: 1, x: 1, y: 0 }

    // Each class has its own fields
    expect(Object.keys(Base._fields)).toEqual(['x', 'y'])
    expect(Object.keys(Cube._fields)).toEqual(['width', 'height'])

    // This method return the proto chain fields
    expect(Base.getFieldNames()).toEqual(['x', 'y'])
    expect(Cube.getFieldNames()).toEqual(['width', 'height', 'x', 'y'])

    // Cube fields handling includes all proto chain fields
    expect(cube).toMatchObject(expected)
    expect(cubeObj).toMatchObject(expected)
  })

  it('should inherit methods', () => {
    class Base extends Schema {
      isBase() {
        return true
      }
    }
    class Cube extends Base {
      isCube() {
        return true
      }
    }
    const cube = new Cube()
    expect(cube.isBase()).toEqual(true)
    expect(cube.isCube()).toEqual(true)
  })

  it.skip('should inherit same resource', () => {
    @resource({ name: 'Base', service: 'foo' })
    class Base extends Schema {}
    class Cube extends Base {}

    expect(Cube._resource).toEqual(Base._resource)
  })
})

describe('Schema virtuals', () => {
  it('should inherit fields from other Schema', () => {
    class Foo extends Schema {
      @field x
      @virtual
      bar = {
        type: String,
        get() {
          return this.x * 2
        },
      }
    }
    const foo = new Foo({ x: 1 })
    expect(foo.bar).toBe(2)
    expect(foo.$toObject()).toMatchObject({ x: 1 })
    expect(foo.$toObject({ virtuals: true })).toMatchObject({ x: 1, bar: '2' })
  })
})

describe('Schema namespace', () => {
  it.skip('should get namespace ', () => {
    @resource({
      name: 'Foo',
      service: 'core',
    })
    class Foo extends Schema {
      @field bar = String
    }
    expect(Foo.ns('bar')).toBe('model.Foo.bar')
  })
})
