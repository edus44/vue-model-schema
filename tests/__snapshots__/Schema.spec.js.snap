// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields properties should collect fields info 1`] = `
Object {
  "filtrable": Array [
    Object {
      "filtrable": true,
      "name": "c",
      "type": Object {
        "name": "Any",
      },
    },
  ],
  "listable": Array [
    Object {
      "listable": true,
      "name": "b",
      "type": Object {
        "name": "Any",
      },
    },
  ],
  "sortable": Array [
    Object {
      "name": "a",
      "sortable": true,
      "type": Object {
        "name": "Any",
      },
    },
  ],
}
`;

exports[`Fields properties should extend fields info 1`] = `
Object {
  "filtrable": Array [],
  "listable": Array [
    Object {
      "listable": true,
      "name": "b",
      "type": Object {
        "name": "Any",
      },
    },
    Object {
      "listable": true,
      "name": "d",
    },
  ],
  "sortable": Array [
    Object {
      "name": "a",
      "sortable": true,
      "type": Object {
        "name": "Any",
      },
    },
  ],
}
`;
